google/cloud/pubsub/integration_tests/publisher_integration_test.cc

```diff
@@ -0,0 +1,111 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsub/admin/topic_admin_client.h"
+#include "google/cloud/pubsub/publisher.h"
+#include "google/cloud/pubsub/testing/random_names.h"
+#include "google/cloud/pubsub/testing/test_retry_policies.h"
+#include "google/cloud/pubsub/version.h"
+#include "google/cloud/credentials.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/internal/random.h"
+#include "google/cloud/testing_util/integration_test.h"
+#include "google/cloud/testing_util/opentelemetry_matchers.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+#include <algorithm>
+
+namespace google {
+namespace cloud {
+namespace pubsub {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::testing_util::IsOk;
+using ::google::cloud::testing_util::StatusIs;
+using ::testing::AnyOf;
+
+class PublisherIntegrationTest
+    : public ::google::cloud::testing_util::IntegrationTest {
+ protected:
+  void SetUp() override {
+    auto project_id =
+        google::cloud::internal::GetEnv("GOOGLE_CLOUD_PROJECT").value_or("");
+    ASSERT_FALSE(project_id.empty());
+    generator_ = google::cloud::internal::DefaultPRNG(std::random_device{}());
+    topic_ = Topic(project_id, pubsub_testing::RandomTopicId(generator_));
+    auto options =
+        Options{}.set<UnifiedCredentialsOption>(MakeGoogleDefaultCredentials());
+    auto const using_emulator =
+        internal::GetEnv("PUBSUB_EMULATOR_HOST").has_value();
+    if (using_emulator) {
+      options = Options{}
+                    .set<UnifiedCredentialsOption>(MakeInsecureCredentials())
+                    .set<internal::UseInsecureChannelOption>(true);
+    }
+
+    auto topic_admin = pubsub_admin::TopicAdminClient(
+        pubsub_admin::MakeTopicAdminConnection(options_));
+
+    auto topic_metadata = topic_admin.CreateTopic(topic_.FullName());
+    ASSERT_THAT(topic_metadata,
+                AnyOf(IsOk(), StatusIs(StatusCode::kAlreadyExists)));
+  }
+
+  void TearDown() override {
+    auto topic_admin = pubsub_admin::TopicAdminClient(
+        pubsub_admin::MakeTopicAdminConnection(options_));
+
+    auto delete_topic = topic_admin.DeleteTopic(topic_.FullName());
+    EXPECT_THAT(delete_topic, AnyOf(IsOk(), StatusIs(StatusCode::kNotFound)));
+  }
+
+  google::cloud::Options options_;
+  google::cloud::internal::DefaultPRNG generator_;
+  Topic topic_ = Topic("unused", "unused");
+};
+
+TEST_F(PublisherIntegrationTest, Basic) {
+  auto publisher = Publisher(MakePublisherConnection(topic_, options_));
+  auto publish =
+      publisher.Publish(MessageBuilder().SetData("test data").Build()).get();
+  ASSERT_STATUS_OK(publish);
+}
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+using ::google::cloud::testing_util::DisableTracing;
+using ::google::cloud::testing_util::EnableTracing;
+
+TEST_F(PublisherIntegrationTest, TracingEnabled) {
```

Comment:
I would install the span catcher in these tests, and set some expectations on the results. While something like #13034 would be nice, it might take some brainpower to design. I think we should start by verifying spans that we know are created all of the time. Something like:

```cc
EXPECT_THAT(span_catcher->GetSpans(), AllOf(
    Contains(SpanNamed("publish")),
    Contains(SpanNamed("google.pubsub.v1.PublisherClient/Publish"))));
```

We can worry about parent/child relationships, sequence of spans, conditional spans (e.g. the backoffs) later.

Reply:
I don't think we want a SpanCatcher here makes sense. That's what the publisher_connection is for.

Reply:
So your test verifies that if you make a client with tracing enabled and use it, the application does not crash.

Why not also verify that if you make a client with tracing enabled and use it, it produces spans?

Reply:
Because that's what the publisher_connection_test does. Then I would have 2 of the same tests.

Reply:
For an end to end test of traces generated, I think there's a better spot then here. Maybe something that creates a publisher, exports the span data to a .txt file, then diffs it.

Reply:
> WDYM by client configuration? Different options?

:+1: 

> What I meant was a solution to #13034

I am trying to understand the solution, but failing to do so.

Reply:
This is just my guess about what happens under the hood in something like https://tracetest.io/